testimony/config.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: UTF-8 -*-
+
+"""
+Parse config for testimony.
+
+Example:
+
+    ---
+    Status:
+        type: choice
+        choices:
+            - Manual
+            - Automated
+    ...
+
+Currently only supported type is 'choice', but in the future more can be added.
+"""
+
+import sys
+import yaml
+
+TOKEN_TYPES = ['choice']
+
+
+def parse_config(filename):
+    """Parse the config."""
+    try:
+        with open(filename, 'r') as fp:
+            data = yaml.load(fp, Loader=yaml.SafeLoader)
+            return {k.lower(): TokenConfig(k, v) for k, v in data.items()}
+    except FileNotFoundError:
+        print('ERROR: File {} not found'.format(filename))
+        sys.exit(1)
+
+
+def update_tokens_dict(tokens_dict, new_keys, default_value=None):
+    """Update dictionary by adding new keys and setting default value."""
+    if isinstance(new_keys, str):
+        new_keys = [token.strip().lower()
+                    for token in new_keys.split(',')
+                    if token.strip()]
+
+    if default_value is None:
+        default_value = {}
+
+    for key in new_keys:
+        if key in tokens_dict:
+            tokens_dict[key].update(default_value)
+        else:
+            tokens_dict[key] = TokenConfig(key, default_value)
+
+
+class TokenConfig(object):
+    """
+    Represent config for one token.
+
+    Curently only checks for value.
+    """
+
+    def __init__(self, name, config):
+        """
+        Initialize token config object.
+
+        Takes name of the token and actual config structure as a param
+        """
+        try:
+            config.get('key')
+        except AttributeError:
+            config = {}
+
+        self.name = name.lower()
+        self.mandatory = config.get('mandatory', False)
+        self.token_type = ''
+
+        if 'type' in config and config['type'] in TOKEN_TYPES:
+            self.token_type = config['type']
+
+        if self.token_type == 'choice':
+            assert 'choices' in config
+            assert isinstance(config['choices'], list)
+            casesensitive = config.get('casesensitive', True)
+            self.choices = [i if casesensitive else i.lower()
+                            for i in config['choices']]
+
+    def update(self, new_values):
+        """Update token configuration with dictionary of new values."""
+        for key, value in new_values.items():
+            setattr(self, key, value)
+
+    def validate(self, what):
+        """Ensure that 'what' meets value validation criteria."""
+        if self.token_type == 'choice':
+            return what.lower() in self.choices
+        return True  # assume valid for unknown types
```

Comment:
If `TokenConfig.mandatory` is True, then this token name will be included in `minimum_tokens` list that is passed in to `DocstringParser`. This way `testcase.has_minimum_tokens` property will be set to False if testcase doesn't contain all required tokens. We check it in `__init__.py:356` and report it under "Test cases missing minimal docstrings".

I can agree that DocstringParser shouldn't be responsible for validating test case metadata tokens, but I tried to not rewrite the whole thing ;) .

Reply:
What I am saying here is to check if the value is not empty, not to check if the minimum set of tokens is available. Does it make sense?

Reply:
I'm not sure if I understand. You mean to add support for config files like that:
```yaml
status:
    required:
```
So, to handle scenarios when `required` key is present, but doesn't have value assigned?

We already do that :) . If key is present, but doesn't have value, then it will be `None`. When we generate `minimum_tokens` list, we cast value of `required` to boolean. So key without value will be False, the same as empty list or dict.